pyccel/parser/semantic.py

```diff
@@ -234,6 +238,16 @@ def annotate(self, **settings):
 
         self._semantic_done = True
 
+        #check if the ast CodeBlock is a Program to add the Garbage collecting
+        cls = (Header, EmptyNode, NewLine, Comment, CommentBlock, Module)
+        body = ast.body
+        is_module = all(isinstance(i,cls) for i in body)
+        if not isinstance(body[-1], Return) and not is_module:
+            dealloc = [Deallocate(i) for i in self._allocs]
+            body += dealloc
+            ast = CodeBlock(body)
+            self._allocs = []
```

Comment:
This does not look like the right place to me. `annotate` is usually only called on the `Module` or the `Program` i.e. the top layer. Normally I think you want to add `Deallocate` in `FunctionDef`.

I wonder if the `CodeBlock` should know what it's parent is?

Reply:
for a Program i think this is the only place I can put it if I don't know the parent from a CodeBlock

Reply:
Ah you are right, a Program is only created in the codegen step! Maybe this could be a function that you can call from `_visit_FunctionDef` too to avoid code duplication?

Reply:
yeah i can make this a function and call it when needed

Reply:
Just the bit that you can use from `_visit_FunctionDef` should be enough :+1: